optuna/_gp/optim_mixed.py

```diff
@@ -0,0 +1,267 @@
+from optuna._gp.search_space import ScaleType
+from optuna.logging import get_logger
+
+
+if TYPE_CHECKING:
+    import scipy.optimize as so
+else:
+    from optuna import _LazyImport
+
+    so = _LazyImport("scipy.optimize")
+
+_logger = get_logger(__name__)
+
+
+def local_search_mixed(
+    acqf_params: AcquisitionFunctionParams,
+    initial_normalized_params: np.ndarray,
+    *,
+    tol: float = 1e-4,
+    max_iter: int = 100,
+) -> tuple[np.ndarray, float]:
+    scale_types = acqf_params.search_space.scale_types
+    bounds = acqf_params.search_space.bounds
+    steps = acqf_params.search_space.steps
+
+    continuous_params = np.where(steps == 0.0)[0]
+
+    inverse_squared_lengthscales = (
+        acqf_params.kernel_params.inverse_squared_lengthscales.detach().numpy()
+    )
+    # This is a technique for speeding up optimization.
+    # We use an isotropic kernel, so scaling the gradient will make
+    # the hessian better-conditioned.
+    continuous_param_scale = 1 / np.sqrt(inverse_squared_lengthscales[continuous_params])
+    continuous_scaled_bounds = [(0, 1 / s) for s in continuous_param_scale]
+
+    noncontinuous_params = np.where(steps > 0)[0]
+    noncontinuous_param_choices = [
+        (
+            np.arange(bounds[i, 1])
+            if scale_types[i] == ScaleType.CATEGORICAL
+            else normalize_one_param(
+                param_value=np.arange(bounds[i, 0], bounds[i, 1] + 0.5 * steps[i], steps[i]),
+                scale_type=ScaleType(scale_types[i]),
+                bounds=(bounds[i, 0], bounds[i, 1]),
+                step=steps[i],
+            )
+        )
+        for i in noncontinuous_params
+    ]
+    noncontinuous_paramwise_xtol = [
+        # Enough to discriminate between two choices.
+        np.min(np.diff(choices), initial=np.inf) / 4
+        for choices in noncontinuous_param_choices
+    ]
+
+    normalized_params = initial_normalized_params.copy()
+    fval = float(eval_acqf_no_grad(acqf_params, normalized_params))
+
+    def optimize_continuous() -> bool:
+        nonlocal normalized_params, fval
+
+        old_continuous_params = normalized_params[continuous_params].copy()
+
+        def negfun_continuous_with_grad(x: np.ndarray) -> tuple[float, np.ndarray]:
+            normalized_params[continuous_params] = x * continuous_param_scale
+            (fval, grad) = eval_acqf_with_grad(acqf_params, normalized_params)
+            # Flip sign because scipy minimizes functions.
+            return (-fval, -grad[continuous_params] * continuous_param_scale)
+
+        x_opt, fval_opt, info = so.fmin_l_bfgs_b(
+            func=negfun_continuous_with_grad,
+            x0=normalized_params[continuous_params] / continuous_param_scale,
+            bounds=continuous_scaled_bounds,
+            pgtol=math.sqrt(tol),
+            maxiter=200,
+        )
+
+        if info["warnflag"] == 2:
+            fval_opt = negfun_continuous_with_grad(x_opt)[0]
+
+        if -fval_opt < fval:
```

Comment:
Opposite?
```suggestion
        if -fval_opt >= fval:
```

Reply:
No, this means "worse" because fval is maximized.

Reply:
Oh sorry, I forgot that this is minimization.